build.rs

```diff
@@ -49,6 +54,168 @@ fn unpack<P: AsRef<Path>>(tgz: impl Read, outdir: P) -> Result<(), std::io::Erro
     Ok(())
 }
 
+// We need the interior mutability of `RefCell` to work around the fact that the
+// `func_macro` callback is not called with a `mut` reference. We need
+// `AssertUnwindSafe` to allow `RefCell` to be used within a `catch_unwind`
+// context inside `bindgen`. We use `Rc` to so that we can get back the `T`
+// after `bindgen` is done with the `Callbacks`, by keeping the reference alive
+// without having to make it `&'static` instead.
+type HideUnwinding<T> = Rc<AssertUnwindSafe<RefCell<T>>>;
+
+struct Callbacks {
+    wrapped: bindgen::CargoCallbacks,
+    state: HideUnwinding<BTreeMap<String, Vec<u8>>>,
+}
+
+impl std::fmt::Debug for Callbacks {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        fmt.debug_struct("Callbacks")
+            .field("wrapped", &self.wrapped)
+            .field("state", &self.state.borrow())
+            .finish()
+    }
+}
+
+impl Callbacks {
+    fn new(bt: HideUnwinding<BTreeMap<String, Vec<u8>>>) -> Self {
+        let wrapped = bindgen::CargoCallbacks;
+        let state = Rc::clone(&bt);
+        Self { wrapped, state }
+    }
+}
+
+impl bindgen::callbacks::ParseCallbacks for Callbacks {
+    fn func_macro(&self, name: &str, value: &[&[u8]]) {
+        self.state.borrow_mut().insert(name.to_owned(), value.join(b" " as &[u8]));
+    }
+
+    fn include_file(&self, filename: &str) {
+        self.wrapped.include_file(filename)
+    }
+}
+
+/// Separates a string into its constituent parts by iteratively removing known
+/// suffixes from the end of the input string.
+// This function is not particularly efficient, and might be reworked.
+fn chop_suffixes(orig: &str) -> Vec<String> {
+    let mut pieces = Vec::new();
+
+    const SUFFIXES: &[&str] = &[
+        "_f", "_d",
+        "_u",
+        "_c", "_i", "_l", "_s",
+        "_uc", "_ui", "_ul", "_us",
+        "_dp", "_fp", "_p", "_pw", "_pwd", "_pwf", "_pww", "_qww",
+        "_w", "_wd", "_wf", "_wp", "_ww", "_wwd", "_wwf", "_www",
+    ];
```

Comment:
There are definitely some ordering restrictions I am not taking advantage of, and there are definitely at least two sets of suffixes, which are mostly non-overlapping (except for some weird exceptions with the internal `mov`s). Let me think about this a bit.

Reply:
@Petelliott, are you all right with requiring a minimum [Rust version of 1.42, in order to support subslice patterns](https://blog.rust-lang.org/2020/03/12/Rust-1.42.html), in case I find them useful ?

Reply:
Never mind, I was already using those subslice patterns in `build.rs`, but I actually want to stop. We can support 1.38.0, it seems.

Reply:
Sorry I didn't see this. I have no issue requiring newer versions of rust.